parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/IndexCacher.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.parquet.hadoop.rewrite;
+
+import org.apache.parquet.column.values.bloomfilter.BloomFilter;
+import org.apache.parquet.hadoop.ParquetFileReader;
+import org.apache.parquet.hadoop.metadata.BlockMetaData;
+import org.apache.parquet.hadoop.metadata.ColumnChunkMetaData;
+import org.apache.parquet.hadoop.metadata.ColumnPath;
+import org.apache.parquet.internal.column.columnindex.ColumnIndex;
+import org.apache.parquet.internal.column.columnindex.OffsetIndex;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A cacher for caching file indexes(ColumnIndex, OffsetIndex, BloomFilter)
+ */
+class IndexCacher {
```

Comment:
@wgtmac Abstracted a cacher class and we could enrich it in the future.

Reply:
Thanks for adding this! Do you think we can move it out of the rewrite module so that the general parquet reader can leverage it in the future?

Reply:
BTW, what about switching it into an interface? We can possibly provide this as a default concrete implementation. A new enum class (e.g. IndexCacheStrategy) for selecting implementations would be beneficial to extensibility.

Reply:
> Thanks for adding this! Do you think we can move it out of the rewrite module so that the general parquet reader can leverage it in the future?

Good idea, this may be useful for file reading. Moved to package `org.apache.parquet.hadoop`

> BTW, what about switching it into an interface? We can possibly provide this as a default concrete implementation. A new enum class (e.g. IndexCacheStrategy) for selecting implementations would be beneficial to extensibility.

Changed to interface and added two strategies: `NONE` and `PRECACHE_BLOCK`.